src/mimeparser.rs

```diff
@@ -217,67 +220,97 @@ impl MimeMessage {
+                    // "3.6 Key Gossip" of <https://autocrypt.org/autocrypt-spec-1.1.0.pdf>
+                    // but only if the mail was correctly signed:
+                    if !signatures.is_empty() {
+                        let gossip_headers =
+                            decrypted_mail.headers.get_all_values("Autocrypt-Gossip");
+                        gossiped_addr =
+                            update_gossip_peerstates(context, message_time, &mail, gossip_headers)
+                                .await?;
+                    }
 
-                        // let known protected headers from the decrypted
-                        // part override the unencrypted top-level
+                    // let known protected headers from the decrypted
+                    // part override the unencrypted top-level
 
-                        // Signature was checked for original From, so we
-                        // do not allow overriding it.
-                        let mut throwaway_from = from.clone();
+                    // Signature was checked for original From, so we
+                    // do not allow overriding it.
+                    let mut signed_from = Vec::new();
 
-                        // We do not want to allow unencrypted subject in encrypted emails because the user might falsely think that the subject is safe.
-                        // See <https://github.com/deltachat/deltachat-core-rust/issues/1790>.
-                        headers.remove("subject");
+                    // We do not want to allow unencrypted subject in encrypted emails because the user might falsely think that the subject is safe.
+                    // See <https://github.com/deltachat/deltachat-core-rust/issues/1790>.
+                    headers.remove("subject");
 
-                        MimeMessage::merge_headers(
-                            context,
-                            &mut headers,
-                            &mut recipients,
-                            &mut throwaway_from,
-                            &mut list_post,
-                            &mut chat_disposition_notification_to,
-                            &decrypted_mail.headers,
-                        );
+                    // TODO question: Why don't we check if `signatures` is empty here?
+                    // If they are empty (and the mail was incorrectly signed), then do we really
+                    // want to allow "secure-join-fingerprint" and "chat-verified" headers?
+                    // (which we didn't allow in the unencrypted headers above)
+                    MimeMessage::merge_headers(
+                        context,
+                        &mut headers,
+                        &mut recipients,
+                        &mut signed_from,
+                        &mut list_post,
+                        &mut chat_disposition_notification_to,
+                        &decrypted_mail.headers,
+                    );
```

Comment:
@link2xt @flub I have a question that's unrelated to AEAP: There are some headers (`secure-join-fingerprint` and `chat-verified`) that are only allowed in the encrypted part. I don't know much about these headers, but it would make sense that they are only allowed if the message is signed correctly? I.e. that we should remove them again if `signatures` is empty.